File: 53.最大子序和.cpp
/*
 * @lc app=leetcode.cn id=53 lang=cpp
 *
 * [53] 最大子序和
 */
#include "vector"
#include "iostream"
#include <limits.h>
using std::vector;
using std::cout;

// @lc code=start
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if(nums.size() == 1) {
            return nums[0];
        }
        int maxSum = nums[0];
        int preMax = maxSum;

        for(int i = 1, len = nums.size(); i < len; i++) {
            preMax = std::max(preMax + nums[i], nums[i]);
            maxSum = std::max(maxSum, preMax);
        }
        return maxSum;
    }
};
// @lc code=end

File: 53.最大子序和.go
/*
 * @lc app=leetcode.cn id=53 lang=golang
 *
 * [53] 最大子序和
 */

// @lc code=start

func maxSubArray(nums []int) int {
	size := len(nums)
	if size == 0 {
		return 0
	}
	var max = func(a, b int) int {
		if a > b {
			return a
		}
		return b
	}
	preM := 0
	m := nums[0]

	for i := 0; i < size; i++ {
		preM = max(preM+nums[i], nums[i])
		m = max(preM, m)
	}
	return m
}
// @lc code=end

